Deconstruction-assignment: nested scenario

jcouv opened the pull request:
Adding support for nested deconstruction-assignments, such as `(x, (y, z)) = M();` or `(x, (y, z)) = (1, (2, 3));`.

The bound node now contains: 
- all the LHS variables (flat list),
- the expression on the right (used as target for deconstruction),
- a list of deconstruction steps (which has a deconstruction method and placeholders for input and outputs, with the method being null in tuple cases), 
- and a list of assignment steps.

Note conversions are still coupled with assignments, but I have a work item (in #11299) to do all conversions first then all assignments.

CC @dotnet/roslyn-compiler for review.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                        Error(diagnostics, ErrorCode.ERR_DeconstructRequiresExpression, node);
+                        return BadExpression(node, checkedVariables.Concat(boundRHS).ToArray());
+                    }
+                }
```
Can this `if` be moved before `try/finally` and `deconstructionSteps` declaration?

jcouv replied:
Yes, indeed. Thanks

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
-            }
+                if (!DeconstructIntoSteps(new BoundDeconstructValuePlaceholder(node.Right, boundRHS.Type), node, diagnostics, checkedVariables, deconstructionSteps, assignmentSteps))
+                {
+                    return new BoundDeconstructionAssignmentOperator(node, checkedVariables, boundRHS, OneOrMany.Create(ImmutableArray<BoundDeconstructionAssignmentOperator.DeconstructStep>.Empty),
```
The `OneOrMany<T>` argument is neither one nor many. Is the `OneOrMany` type misnamed or should the argument in this case be simply `ImmutableArray<T>`?

cston commented on src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs:
```diff
@@ -105,6 +105,44 @@ public T Current
+
+            for (int i = 0; i < Count; i++)
+            {
+                if (this[i].Equals(other[i]))
```
`!`?

cston commented on src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs:
```diff
@@ -105,6 +105,44 @@ public T Current
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
```
`Add` needs to be changed to use `_one` rather than `_many` when adding to an empty set. Otherwise, `base.GetHashCode()` will return different values for the `_one` and `_many.Length == 1` cases, even though those will compare `Equals`. We may be better off not allowing an empty `OneOrMany<T>`.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -1343,5 +1354,324 @@ public void Deconstruct(out int a, out int b)
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello", parseOptions: TestOptions.Regular.WithTuplesFeature());
```
This should return "2 hello" instead. I'll fix.

VSadov replied:
Why would it be "2 hello" ?

jcouv replied:
Oops, sorry, I meant in the following test. `AssignWithPostfixOperator`

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -280,11 +280,18 @@ public override Symbol ExpressionSymbol
+            public MethodSymbol DeconstructMemberOpt; // the deconstruct member, or null if tuple deconstruction
+            public BoundDeconstructValuePlaceholder TargetPlaceholder;
+            public ImmutableArray<BoundDeconstructValuePlaceholder> OutputPlaceholders; // placeholders for the outputs produced by this deconstruction
         }
```
Consider making these two types immutable.

cston replied:
Can these types be `struct`s?

jcouv replied:
Yes, I thought about `struct` too. I have a note on that #11299. I think I'll just make the change.

Also, from our discussion, I will move back from `OneOrMany` to simply use `ImmutableArray`.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+            {
+                DeconstructMemberOpt = null,
+                TargetPlaceholder = targetPlaceholder,
+                OutputPlaceholders = tupleTypes.SelectAsArray(type => new BoundDeconstructValuePlaceholder(syntax, type))
```
Consider using ```tupleTypes.SelectAsArray((t, s) => new BoundDeconstructValuePlaceholder(s, t), syntax)``` to avoid allocating a closure class. Similar comment for `DeconstructWithDeconstruct`.

jcouv replied:
Thanks. Done

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+            if (isFlat)
+            {
+                return variables;
+            }
```
Perhaps ```if (variables.All(v => v.Kind == BoundKind.DeconstructionVariables))```.

jcouv replied:
Done. I was nervous about using LINQ, but now I understand the relevant factor (does it capture local variables?).

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+            return builder.ToImmutableAndFree();
+        }
+
+        private void FlattenDeconstructVariables(ImmutableArray<BoundExpression> variables, ArrayBuilder<BoundExpression> builder)
```
`static` for both overloads.

jcouv replied:
Thanks. Done

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
 
-            TypeSymbol voidType = GetSpecialType(SpecialType.System_Void, diagnostics, node);
-            return new BoundDeconstructionAssignmentOperator(node, checkedVariables, typedRHS, deconstructMemberOpt: null, assignments: assignments, type: voidType);
+            return true;
```
```return DeconstructOrAssignOutputs(...);```

jcouv replied:
Thanks. Done

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                return false;
+            }
+
+            return true;
```
Is it just the first step that is different between `DeconstructTuple` and `DeconstructWithDeconstruct`? If so, could these methods contain that step only?

jcouv replied:
Yes, I can indeed push the code factoring farther. Thanks!
I refactored as you suggested, but changed the `DeconstructionStep` type back to a class (so that I can use null for error cases).

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                        ArrayBuilder<BoundDeconstructionAssignmentOperator.AssignmentStep> assignmentSteps)
         {
             // symbol and parameters for Deconstruct
             DiagnosticBag bag = new DiagnosticBag();
```
`DiagnosticBag.GetInstance()`

jcouv replied:
Done

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+
+                if (variable.Kind == BoundKind.DeconstructionVariables)
+                {
+                    var nestedVariables = ((BoundDeconstructionVariables)variables[i]).Variables;
```
`variable`

cston commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -394,13 +392,21 @@
+    <Field Name="AssignmentSteps" Type="ImmutableArray&lt;BoundDeconstructionAssignmentOperator.AssignmentStep&gt;" Null="NotApplicable" SkipInVisitor="true"/>
+  </Node>
+
+  <Node Name="BoundDeconstructionVariables" Base="BoundExpression">
```
Why a `BoundExpression` rather than `BoundNode`?

cston replied:
Do we need a `BoundNode` at all or could we pass around `ImmutableArray<BoundExpression>` instead?

jcouv replied:
So, this bound node is never output from binding at all. But during binding, I have bound expressions for the LHS, which need to be organized in a tree.
Because all simple variables are `BoundExpression`, it is convenient that this is also `BoundExpression` (rather than `BoundNode`). But I'm open to suggestions. 
I agree that having a bound node that never escapes binding is strange.

cston replied:
Where do we use `checkedVariables` as (array of) `BoundExpression` rather than `BoundNode`? Or is it more than just `checkedVariables` that is the concern?

AlekseyTs replied:
Given the current implementation, it looks like BoundDeconstructionVariables never escapes the binder. This is just a helper node, used in the process of binding, never gets stored in the tree produced by the Binder. Please provide a comment reflecting that.

jcouv replied:
Removed `BoundDeconstructionVariables` in favor of a simple tree structure (which is not a bound node):
``` C#
        private class DeconstructionVariable
        {
            public BoundExpression Single;
            public ImmutableArray<DeconstructionVariable>? Nested;
            ...
        }
```

cston commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -1049,6 +1034,21 @@ public override void Accept(OperationVisitor visitor)
+
+        public override void Accept(OperationVisitor visitor)
+        {
+            visitor.VisitNoneOperation(this);
```
Why not visit `Variables`?

jcouv replied:
This is a dummy implementation, like `BoundTupleExpression`.
My understanding is that this implement IOperation, which John Hamby would take care of.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -10,6 +10,48 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
     [CompilerTrait(CompilerFeature.Tuples)]
     public class CodeGenDeconstructTests : CSharpTestBase
     {
+        const string commonSource =
```
`commonSource` is only used in one test. Consider inlining in the test, for readability.

jcouv replied:
I expect I'll use this common code in more tests.

jcouv replied:
I now have a second test using it. I'll probably have more usages soon.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,293 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
-            MethodSymbol deconstructMethod = FindDeconstruct(checkedVariables, boundRHS, node, bag);
+            var bag = DiagnosticBag.GetInstance();
+            MethodSymbol deconstructMethod = FindDeconstruct(variables.Length, targetPlaceholder, syntax, bag);
             if (!diagnostics.HasAnyErrors())
```
Is `bag` only added if `diagnostics` is empty to avoid cascading errors? Are we sure existing diagnostics are from the same expression (in other words, are we sure `bag` contains errors related to those in `diagnostics`)? Perhaps add `bag` always.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,293 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+        /// <summary>
+        /// Takes the outputs from the previous deconstructionStep and depending on the structure of variables, will generate further deconstructions, or simply assignments.
+        /// </summary>
+        public bool DeconstructOrAssignOutputs(
```
`private`?

jcouv replied:
Yes. Fixed.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,293 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                {
+                    var assignment = MakeAssignmentInfo(variable, deconstructionStep.OutputPlaceholders[i].Type, deconstructionStep.OutputPlaceholders[i], syntax, diagnostics);
+                    assignmentSteps.Add(assignment);
+                }
```
Perhaps extract variable for `deconstructionStep.OutputPlaceholders[i]`.

jcouv replied:
Done.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,293 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+        }
+
+        /// <summary>
+        /// For cases where the RHS of a deconstruction-assignment has not type (TupleLiteral), we squint and look at the LHS as a tuple type to give the RHS a type.
```
`has no type`?

jcouv replied:
Fixed.

cston commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs:
```diff
@@ -1721,14 +1721,40 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
+            {
+                if (variable.Kind == BoundKind.DeconstructionVariables)
+                {
+                    VisitDeconstructionVariables((BoundDeconstructionVariables)variable);
```
Doesn't `base.VisitDeconstructionAssignment` do this? Same comment for `VisitDeconstructionVariables` below.

cston replied:
If the `Assign` call is the only difference from the `base` call, consider adding a `virtual VisitDeconstructionAssignmentLeftVariable` method to `PreciseAbstractFlowPass` and override that method here instead. Similarly for `VisitDeconstructionVariables`.

jcouv replied:
So, I removed `BoundDeconstructionVariables`, which simplifies and removes a bunch of data flow code. I'm leaving the rest as-is with `PROTOTYPE(tuples)`.
The `Assign` call is a stub, not the full implementation that I intend.

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -12,76 +11,91 @@ internal sealed partial class LocalRewriter
             BoundExpression loweredRight = VisitExpression(node.Right);
-            ImmutableArray<BoundExpression> rhsValues;
+            AddPlaceholderReplacement(node.DeconstructSteps[0].TargetPlaceholder, loweredRight);
+            placeholders.Add(node.DeconstructSteps[0].TargetPlaceholder);
```
Is this `AddPlaceholderReplacement` case also handled by the `foreach` loop below?

jcouv replied:
I'm not sure I understand the question.
Here we bootstrap the first placeholder for the RHS value. After the RHS value is deconstructed, it will populate proper placeholders for subsequence deconstructs, and so on. Those same placeholders are also referenced by the assignment step.
At the very end, we clear all those "carried-through" placeholders.

cston replied:
My mistake, I didn't realize `AccessTupleFields` and `CallDeconstruct` drilled in to fields or parameters of each step.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -394,13 +392,21 @@
     <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
 
+    <!-- Various assignable expressions, including some BoundDeconstructionVariables for nested variables -->
     <Field Name="LeftVariables" Type="ImmutableArray&lt;BoundExpression&gt;"/>
```
Why do we need BoundDeconstructionVariables in this list?

AlekseyTs replied:
Presence of these node can potentially confuse SemanticModel, etc.

---
In reply to: [65776149](https://github.com/dotnet/roslyn/pull/11715#discussion_r65776149) [](ancestors = 65776149)

AlekseyTs replied:
Ok, given the current implementation, it looks like LeftVariable is always a flat list of targets, BoundDeconstructionVariables is never among them. Please adjust the comment accordingly. 

---
In reply to: [65778993](https://github.com/dotnet/roslyn/pull/11715#discussion_r65778993) [](ancestors = 65778993,65776149)

jcouv replied:
@AlekseyTs, from discussion with Chuck I will remove this node and store the tree of variables some other way (without creating new kind of bound node).

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -280,11 +280,18 @@ public override Symbol ExpressionSymbol
     internal sealed partial class BoundDeconstructionAssignmentOperator : BoundExpression
     {
-        internal class AssignmentInfo
+        internal class AssignmentStep
```
I think AssignmentStep and DeconstructStep should be regular bound nodes

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs:
```diff
@@ -280,11 +280,18 @@ public override Symbol ExpressionSymbol
+        internal class DeconstructStep
+        {
+            public MethodSymbol DeconstructMemberOpt; // the deconstruct member, or null if tuple deconstruction
+            public BoundDeconstructValuePlaceholder TargetPlaceholder;
```
I would make this BoundExpression and put Right directly as the target for the first deconstruct, removing it from the parent node.

jcouv replied:
> put Right directly as the target for the first deconstruct, removing it from the parent node.

I didn't get that part.

AlekseyTs replied:
Never mind, since we might not create the deconstruct step in case of an error, the suggestion is not an option.

---
In reply to: [65796074](https://github.com/dotnet/roslyn/pull/11715#discussion_r65796074) [](ancestors = 65796074)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                else
+                {
+                    // expression without type such as `null`
+                    Error(diagnostics, ErrorCode.ERR_DeconstructRequiresExpression, node);
```
Should we explicitly call node.Display rather than passing the raw node?

jcouv replied:
The third parameter of `Error` takes a `CSharpSyntaxNode`. So I think this is correct.

AlekseyTs replied:
Got that now. It is used for location and not as an argument of the message.

---
In reply to: [65793066](https://github.com/dotnet/roslyn/pull/11715#discussion_r65793066) [](ancestors = 65793066)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                {
+                    // expression without type such as `null`
+                    Error(diagnostics, ErrorCode.ERR_DeconstructRequiresExpression, node);
+                    return BadExpression(node, checkedVariables.Concat(boundRHS).ToArray());
```
Should FlattenDeconstructVariables

jcouv replied:
Good catch. Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+            var assignmentSteps = ArrayBuilder<AssignmentStep>.GetInstance(1);
+            try
+            {
+                if (!DeconstructIntoSteps(new BoundDeconstructValuePlaceholder(node.Right, boundRHS.Type), node, diagnostics, checkedVariables, deconstructionSteps, assignmentSteps))
```
I think that even when some deconstruct steps are failing we should still preserve what we managed to figure out and proceed on binding assignments as much as we can figure out, probably suppressing additional diagnostics. This will likely to benefit SemanticModel and IOperation. We would want to expose as much information as we are able to figure out, even for error scenarios.

jcouv replied:
Yes. I think switching `DeconstructionStep` and `AssignmentStep` to be bound nodes will help with this, as we'll be able to capture all the ones computed so far into the produced error bound node.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+        /// </summary>
+        static private DeconstructStep MakeTupleDeconstructStep(
+                        BoundDeconstructValuePlaceholder targetPlaceholder,
+                        AssignmentExpressionSyntax syntax,
```
It feels like we would not want to associate all nodes and errors with the original AssignmentExpressionSyntax. I think we should gradually move from the syntax of the left target to its constituent parts that are the targets of each particular deconstruct.

AlekseyTs replied:
The syntax could easily be tracked through syntax that BoundDeconstructionVariables nodes are associated with.

---
In reply to: [65782761](https://github.com/dotnet/roslyn/pull/11715#discussion_r65782761) [](ancestors = 65782761)

jcouv replied:
I agree. I actually have a work item in #11299 to better track locations in error messages.
I'll give it a shot. My thinking is that I'll need to track a left syntax and a right syntax, and we should get specific squiggles on both sides if possible.

jcouv replied:
After having removed `BoundDeconstructionVariables`, this becomes a bit more difficult. I'll keep the work-item open for later PR.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
             }
 
-            return BindDeconstructWithDeconstruct(node, diagnostics, boundRHS, checkedVariables);
+            deconstructionSteps.Add(step);
```
Probably shouldn't add null into the builder.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+        /// </summary>
+        static private DeconstructStep MakeNonTupleDeconstructStep(
+                        BoundDeconstructValuePlaceholder targetPlaceholder,
+                        AssignmentExpressionSyntax syntax,
```
Same comment about associated syntax node.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
-            MethodSymbol deconstructMethod = FindDeconstruct(checkedVariables, boundRHS, node, bag);
+            var bag = DiagnosticBag.GetInstance();
+            MethodSymbol deconstructMethod = FindDeconstruct(variables.Length, targetPlaceholder, syntax, bag);
             if (!diagnostics.HasAnyErrors())
```
I think we shouldn't be suppressing diagnostics based on whether there are other errors in the bag, unless we are controlling what might be in the bag. We can suppress additional diagnostics based on the fact that we reported something already, that was definitely related to the things we are going to be doing next.

jcouv replied:
Makes sense. I removed the local bag and just use `diagnostics`.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1726,294 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+
+                    if (!DeconstructIntoSteps(valuePlaceholder, syntax, diagnostics, nestedVariables, deconstructionSteps, assignmentSteps))
+                    {
+                        return false;
```
Should record the fact of the failure and continue with the rest, we could still figure out more and/or report more useful diagnostics for other parts.

jcouv replied:
Done. I'm now collecting all the deconstruction steps that I can, even after encountering an error.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs:
```diff
@@ -1526,6 +1526,12 @@ public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstruct
+        public override BoundNode VisitDeconstructionVariables(BoundDeconstructionVariables node)
+        {
+            Debug.Assert(false, "BoundDeconstructionVariables should not make it past binding stage.");
+            return null;
```
I think it is more appropriate to throw Unreachable here.

AlekseyTs replied:
And seal the method.

---
In reply to: [65786461](https://github.com/dotnet/roslyn/pull/11715#discussion_r65786461) [](ancestors = 65786461)

jcouv replied:
Removed `BoundDeconstructionVariables`.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs:
```diff
@@ -287,6 +282,17 @@ private void RemovePlaceholderReplacement(BoundValuePlaceholderBase placeholder)
+        /// <summary>
+        /// Remove all the listed placeholders.
+        /// </summary>
+        private void RemovePlaceholderReplacements(IEnumerable<BoundValuePlaceholderBase> placeholders)
```
Could we use more specific type rather than IEnumerable<BoundValuePlaceholderBase>? ArrayBuilder/ImmutableArray

cston replied:
Perhaps inline the method since it is only used in one place.

jcouv replied:
I suspect in the future it will be handy again to be able to remove multiple placeholders. I'm not sure what is the downside of using `IEnumerable` to keep options open.

jcouv replied:
@AlekseyTs Why not `IEnumerable`?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -12,76 +11,91 @@ internal sealed partial class LocalRewriter
+        /// <summary>
+        /// Adds the side effects to stores and returns temporaries (as a flat list) to access them.
+        /// </summary>
+        private ImmutableArray<BoundExpression> LhsSideEffects(ImmutableArray<BoundExpression> variables, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> stores)
```
Consider spelling out what 'Lhs' stands for.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -12,76 +11,91 @@ internal sealed partial class LocalRewriter
+
+            foreach (var variable in variables)
+            {
+                lhsReceivers.Add(TransformCompoundAssignmentLHS(variable, stores, temps, isDynamicAssignment: false));
```
Not related to this change, but are we certain that isDynamicAssignment should always be false?

jcouv replied:
I'm not sure. Added a follow-up item "Verify correct handling of dynamic" to #11299 and a `PROTOTYPE` marker.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs:
```diff
@@ -12,76 +11,91 @@ internal sealed partial class LocalRewriter
-                lhsReceivers.Add(TransformCompoundAssignmentLHS(variable, stores, temps, isDynamicAssignment: false));
-            }
+            // evaluate left-hand-side side-effects
+            var lhsTemps = LhsSideEffects(node.LeftVariables, temps, stores);
```
These are not temps, these are targets of the assignments. Correct?

jcouv replied:
Correct. I'll rename to `lhsTargets`.

AlekseyTs replied:
Still not renamed.

---
In reply to: [65793459](https://github.com/dotnet/roslyn/pull/11715#discussion_r65793459) [](ancestors = 65793459)

jcouv replied:
Something is messed up with git. My local file doesn't show this anymore, yet `git status` is clean and `git blame` shows the wrong content...

jcouv replied:
Seems I'm looking at the wrong enlistment. Woke up on the wrong foot ;-)
I'll fix the name. Sorry about that.

AlekseyTs commented:
Done with the review pass

jcouv commented:
Pushed some commits addressing the feedback.
I'm on the fence about removing `BoundDeconstructionVariables` as it would make it easier to track portions of syntax on the LHS.

AlekseyTs commented:
Having BoundDeconstructionVariables as a helper node seems fine as long as it doesn't survive initial binding. It does provide an easy way to track relevant syntax node. Otherwise we would have to use a tuple for this. 

---
In reply to: [223733070](https://github.com/dotnet/roslyn/pull/11715#issuecomment-223733070) [](ancestors = 223733070)

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,283 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+            return new BoundDeconstructionDeconstructStep(syntax, deconstructMethod, targetPlaceholder, deconstructMethod.Parameters.SelectAsArray((p, s) => new BoundDeconstructValuePlaceholder(s, p.Type), syntax));
+        }
+
+        private class DeconstructionVariable
```
Perhaps `struct`.

cston replied:
Perhaps `struct` with `readonly` fields.

jcouv replied:
I had it as a `struct` initially, but then the type couldn't be loaded. I'm not 100% sure why, especially given the presence of nullable.

jcouv replied:
I'll leave as `class` since working around the CLR bugs ripples too far in the code.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,283 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+        private class DeconstructionVariable
+        {
+            public BoundExpression Single;
+            public ImmutableArray<DeconstructionVariable>? Nested;
```
Consider using non-nullable `ImmutableArray<DeconstructionVariable>` where the unset value is `default(Immutable<DeconstructionVariable>)`.

VSadov replied:
There is a type loader bug that would cause a struct S1 containing ImmutableArray<S1> be rejected as having a layout cycle (there is not real cycle though since IA wraps a ref array)
 CLR knows about this bug, but it is too risky area to fix right now.

jcouv replied:
Done.

jcouv commented:
@AlekseyTs @cston I pushed a change that uses `default` instead of a nullable and avoids `IEnumerable`.
The tracking of more precise locations is deferred to a follow-up work item.

jcouv commented:
@AlekseyTs @cston Please let me know if ok to merge. Thanks

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,283 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
-            TypeSymbol lhsAsTuple = TupleTypeSymbol.Create(locationOpt: null, elementTypes: lhsTypes, elementLocations: default(ImmutableArray<Location>), elementNames: default(ImmutableArray<string>), compilation: Compilation, diagnostics: diagnostics);
-            var typedRHS = GenerateConversionForAssignment(lhsAsTuple, boundRHS, diagnostics);
+            public BoundExpression Single;
+            public ImmutableArray<DeconstructionVariable> Nested;
```
`readonly` for both fields.

jcouv replied:
Done

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -1724,104 +1724,283 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
 
-            TypeSymbol voidType = GetSpecialType(SpecialType.System_Void, diagnostics, node);
-            return new BoundDeconstructionAssignmentOperator(node, checkedVariables, typedRHS, deconstructMemberOpt: null, assignments: assignments, type: voidType);
+            public bool IsNested => Nested != default(ImmutableArray<DeconstructionVariable>);
```
```=> !Nested.IsDefault```

jcouv replied:
Thanks. Fixed

cston commented:
LGTM

AlekseyTs commented:
LGTM, modulo the fact that ```lhsTemps``` are still not renamed to ```lhsTemps``` in LocalRewriter.

jcouv commented:
Merged. Thanks for the review and feedback.
